rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
+        DataType::List(_) => Arc::new(ListArray::from(data)) as ArrayRef,
+        DataType::Struct(_) => Arc::new(StructArray::from(data)) as ArrayRef,
+        dt => panic!("Unexpected data type {:?}", dt),
     }
+}
 
-    pub fn null_count(&self) -> usize {
-        self.null_count
-    }
+/// ----------------------------------------------------------------------------
+/// Implementations of different array types
 
-    pub fn validity_bitmap(&self) -> &Option<Bitmap> {
-        &self.validity_bitmap
-    }
+struct RawPtrBox<T> {
+    inner: *const T,
+}
 
-    pub fn get(&self, i: usize) -> &[T] {
-        self.data.get(i)
+impl<T> RawPtrBox<T> {
+    fn new(inner: *const T) -> Self {
+        Self { inner: inner }
     }
 
-    pub fn list(&self) -> &List<T> {
-        &self.data
+    fn get(&self) -> *const T {
+        self.inner
     }
 }
 
-/// Create a ListArray<T> from a List<T> without null values
-impl<T> From<List<T>> for ListArray<T>
-where
-    T: ArrowPrimitiveType,
-{
-    fn from(list: List<T>) -> Self {
-        let len = list.len();
-        ListArray {
-            len,
-            data: list,
-            validity_bitmap: None,
-            null_count: 0,
-        }
-    }
+unsafe impl<T> Send for RawPtrBox<T> {}
+unsafe impl<T> Sync for RawPtrBox<T> {}
+
+/// Array whose elements are of primitive types.
+pub struct PrimitiveArray<T: ArrowPrimitiveType> {
+    data: ArrayDataRef,
+    /// Pointer to the value array. The lifetime of this must be <= to the value buffer
+    /// stored in `data`, so it's safe to store.
+    raw_values: RawPtrBox<u8>,
+    _phantom: PhantomData<T>,
 }
 
-/// Create ListArray<u8> from Vec<&'static str>
-impl From<Vec<&'static str>> for ListArray<u8> {
-    fn from(v: Vec<&'static str>) -> Self {
-        let mut builder: ListBuilder<u8> = ListBuilder::with_capacity(v.len());
-        for s in v {
-            builder.push(s.as_bytes())
+/// Macro to define primitive arrays for different data types and native types.
+macro_rules! def_primitive_array {
+    ($data_ty:path, $native_ty:ident) => {
+        impl PrimitiveArray<$native_ty> {
+            pub fn new(length: i64, values: Buffer, null_count: i64, offset: i64) -> Self {
+                let array_data = ArrayData::builder($data_ty)
+                    .length(length)
+                    .add_buffer(values)
+                    .null_count(null_count)
+                    .offset(offset)
+                    .build();
+                PrimitiveArray::from(array_data)
+            }
+
+            /// Returns a `Buffer` holds all the values of this array.
+            pub fn values(&self) -> Buffer {
+                self.data.buffers()[0].copy()
+            }
+
+            /// Returns a raw pointer to the values of this array.
+            pub fn raw_values(&self) -> *const $native_ty {
+                unsafe { mem::transmute(self.raw_values.get().offset(self.data.offset() as isize)) }
```

Comment:
I'm curious, why is `mem::transmute` needed here?

Reply:
This is because `raw_values` is a `*const u8` and we need to cast it to `*const $native_ty`, which is what `mem::transmute` does.

Reply:
but this would require the `*const u8` is aligned correctly. I think we should check this here. Otherwise, you end up with undefined behavior (or on x86, at least with a SIGBUS).